core/src/main/java/hudson/model/labels/LabelExpression.java

```diff
@@ -210,4 +225,158 @@ public LabelOperatorPrecedence precedence() {
+        public abstract FormValidation check(@NonNull Job<?, ?> job, @NonNull Label label);
+
+    }
+
+    /**
+     * Generates auto-completion candidates for a (partial) label.
+     *
+     * @param label The (partial) label for which auto-completion is being requested.
+     * @return A set of auto-completion candidates.
+     * @since TODO
+     */
+    public static AutoCompletionCandidates autoComplete(@Nullable String label) {
+        AutoCompletionCandidates c = new AutoCompletionCandidates();
+        Set<Label> labels = Jenkins.get().getLabels();
+        List<String> queries = new AutoCompleteSeeder(label).getSeeds();
+
+        for (String term : queries) {
+            for (Label l : labels) {
+                if (l.getName().startsWith(term)) {
+                    c.add(l.getName());
+                }
+            }
+        }
+        return c;
+    }
+
+    /**
+     * Validates a label expression.
+     *
+     * @param expression The expression to validate.
+     * @return The validation result.
+     * @since TODO
+     */
+    @NonNull
+    public static FormValidation validate(@Nullable String expression) {
+        return LabelExpression.validate(expression, null);
+    }
+
+    /**
+     * Validates a label expression.
+     *
+     * @param expression The label expression to validate.
+     * @param job        The job context, if applicable; used for potential additional restrictions.
+     * @return The validation result.
+     * @since TODO
+     */
+    // FIXME: Should the messages be moved, or kept where they are for backward compatibility?
+    @NonNull
+    public static FormValidation validate(@Nullable String expression, @CheckForNull Job<?, ?> job) {
+        if (Util.fixEmpty(expression) == null)
+            return FormValidation.ok(); // nothing typed yet
+        try {
+            Label.parseExpression(expression);
+        } catch (ANTLRException e) {
+            return FormValidation.error(e,
+                    Messages.AbstractProject_AssignedLabelString_InvalidBooleanExpression(e.getMessage()));
+        }
+        Jenkins j = Jenkins.get();
+        Label l = j.getLabel(expression);
+        if (l.isEmpty()) {
+            for (LabelAtom a : l.listAtoms()) {
+                if (a.isEmpty()) {
+                    LabelAtom nearest = LabelAtom.findNearest(a.getName());
+                    return FormValidation.warning(Messages.AbstractProject_AssignedLabelString_NoMatch_DidYouMean(a.getName(),nearest.getDisplayName()));
+                }
+            }
+            return FormValidation.warning(Messages.AbstractProject_AssignedLabelString_NoMatch());
+        }
+        if (job != null) {
+            if (job instanceof AbstractProject) { // Use any project-oriented label validators
+                final AbstractProject<?, ?> project = (AbstractProject<?,?>) job;
+                for (AbstractProject.LabelValidator v : j.getExtensionList(AbstractProject.LabelValidator.class)) {
```

Comment:
or can use `ExtensionList.lookup`

Reply:
I just kept the original code (other than wrapping it in a is-it-a-project check); I can look at using this other method instead.

Reply:
> I just kept the original code (other than wrapping it in a is-it-a-project check)

Fine, this was just FYI.